### DISCUSSION TITLE
"greatest-n-per-group" query howto

### DISCUSSION THREAD

**Discussion by trueqbit** [fnc12/sqlite_orm#773] [Q&A]
Is it possible to perform a "greatest-n-per-group" query? I guess this belongs to one of the most needed tasks...

In other words: having a table consisting of multiple results for items, I want to select a single result row per item, based on a condition like an aggregated value.

This is possible using a (self) join and filtering by aggregated value (in case it is unique), or using window functions.
Especially because I need a 'self join' I can't figure out how to do this using sqlite_orm.

My specific example, given this table:
rowid|itemid|some_date|flag
-----|------|---------|----

Using a (self) join:
```sql
select count(r.itemid)
from some_result r
inner join (
  select itemid,
    max(some_date) as max_date
  from some_result
  group by itemid
  ) wnd
  on wnd.itemid = r.itemid
    and r.some_date = wnd.max_date
-- other conditions
where r.flag = 1
```

Using window functions:
```sql
select count(itemid)
from (
  select itemid,
    flag,
    row_number() over (
      partition by itemid order by some_date desc, rowid desc
      ) as rank
  from some_result
  ) as wnd
where wnd.rank = 1
  -- other conditions
  and wnd.flag = 1
```

Also see those answers on SO:
[SQL select only rows with max value on a column (traditional self join)](https://stackoverflow.com/a/7745635/279251)
[SQL select only rows with max value on a column (window function)](https://stackoverflow.com/a/38854846/279251)
[Select info from table where row has max date](https://stackoverflow.com/a/19433107/279251)

**Comment by fnc12** [comment]
Hello @trueqbit . Unfortunately such queries cannot be implemented using `sqlite_orm` for now. The first needs `INNER JOIN` from `SELECT` feature implementation. The second needs window functions support and `FROM` from `SELECT` feature support. 
To implement `INNER JOIN` from `SELECT` and `FROM` from `SELECT` we need to create a clear and intuitive API for this. If you have any ideas please share

**Comment by trueqbit** [comment]
Using CTEs it is possible to hoist the `INNER JOIN` with a subselect into a CTE:

```sql
with wnd(itemid,
         max_date) as
  (select itemid,
          max(some_date)
   from some_result
   group by itemid)
select count(r.itemid)
from some_result r
inner join wnd on (r.itemid = wnd.itemid)
  and (r.some_date = wnd.max_date)
-- other conditions
where (r.flag = 1)
```

See the dedicated `greatest_n_per_group` [example](https://github.com/fnc12/sqlite_orm/blob/69917bd09a84970881d755828d60d8edf79728c6/examples/common_table_expressions.cpp#L1031).
